src/Common/FileSegment.cpp

```diff
@@ -0,0 +1,443 @@
+#include "FileSegment.h"
+#include <base/getThreadId.h>
+#include <Common/FileCache.h>
+#include <Common/hex.h>
+#include <IO/WriteBufferFromString.h>
+#include <IO/Operators.h>
+#include <filesystem>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int REMOTE_FS_OBJECT_CACHE_ERROR;
+    extern const int LOGICAL_ERROR;
+}
+
+FileSegment::FileSegment(
+        size_t offset_,
+        size_t size_,
+        const Key & key_,
+        IFileCache * cache_,
+        State download_state_)
+    : segment_range(offset_, offset_ + size_ - 1)
+    , download_state(download_state_)
+    , file_key(key_)
+    , cache(cache_)
+#ifndef NDEBUG
+    , log(&Poco::Logger::get(fmt::format("FileSegment({}) : {}", getHexUIntLowercase(key_), range().toString())))
+#else
+    , log(&Poco::Logger::get("FileSegment"))
+#endif
+{
+    if (download_state == State::DOWNLOADED)
+        reserved_size = downloaded_size = size_;
+    else if (download_state != State::EMPTY)
+        throw Exception(ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR, "Can create cell with either DOWNLOADED or EMPTY state");
+}
+
+FileSegment::State FileSegment::state() const
+{
+    std::lock_guard segment_lock(mutex);
+    return download_state;
+}
+
+size_t FileSegment::getDownloadOffset() const
+{
+    std::lock_guard segment_lock(mutex);
+    return range().left + downloaded_size;
+}
+
+String FileSegment::getCallerId()
+{
+    return getCallerIdImpl(false);
+}
+
+String FileSegment::getCallerIdImpl(bool allow_non_strict_checking)
+{
+    if (IFileCache::shouldBypassCache())
+    {
+        /// getCallerId() can be called from completeImpl(), which can be called from complete().
+        /// complete() is called from destructor of CachedReadBufferFromRemoteFS when there is no query id anymore.
+        /// Allow non strict checking in this case. This works correctly as if getCallerIdImpl() is called from destructor,
+        /// then we know that caller is not a downloader, because downloader is reset each nextImpl() call either
+        /// manually or via SCOPE_EXIT.
+
+        if (allow_non_strict_checking)
+            return "None";
+
+        throw Exception(ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR, "Cannot use cache without query id");
```

Comment:
Yeah, thanks, I will change in those places.
Initially I started making it REMOTE_FS_OBJECT_CACHE_ERROR instead of LOGICAL_ERROR, because I wanted to check certain errors for EXPECT_THROW in unit tests, but then it would not work in debug build unit tests, so I started to not use logical error...